### PR TITLE
Refactoring the insert activity controller to make use of queue

### PR DESCRIPTION
**Pull request by C0SMlC**

### CODE REVIEW

File: src/controllers/userActivites.controller/insertUserActivity/index.js [hunk 1]
```diff
@@ -31,15 +31,12 @@ const getUserPoints = catchAsync(async (req, res) => {
...
   const userId = getUserIdQueryResponse.user[0].id;
 
-  const insertUserActivityVariables = {
-    userId,
-    activityId,
-    direction,
-    status: 'unprocessed',
-  };
-  const insertUserActivityResponse = await Hasura(insertUserActivityQuery, insertUserActivityVariables);
+  insertUserActivity(activityIdentifier, 'positive', userId, [entityId]);
```
Comment (Shivz3232) [inline src/controllers/userActivites.controller/insertUserActivity/index.js]: Let's include direction also to be sent from the frontend

Comment (Shivz3232) [inline src/controllers/userActivites.controller/insertUserActivity/index.js]: Can we validate entity id? For example for share project activity what is they send a project is that doesn't exist?

**Review by Shivz3232 — CHANGES_REQUESTED** [review]

File: src/controllers/userActivites.controller/insertUserActivity/index.js [hunk 2]
```diff
@@ -29,17 +30,24 @@ const getUserPoints = catchAsync(async (req, res) => {
     });
   }
 
+  const isEntityValid = (await validateEntity(activityIdentifier, entityId)) || false;
+
+  if (!isEntityValid) {
+    return res.status(404).json({
```
Comment (Shivz3232) [inline src/controllers/userActivites.controller/insertUserActivity/index.js]: Status should be 400

File: src/utils/validateEntity/index.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+const { query: Hasura } = require('../hasura');
+const { findIdea, findProject, findThought } = require('./queries/queries');
+
+const identifierMap = {
```
Comment (Shivz3232) [inline src/utils/validateEntity/index.js:4]: Can we switch to using DB & maintaining a cache just like we are doing in [getActivityId](https://github.com/Shivz3232/inovact-nodejs-veyga/blob/409034ef7df2d8063d6c36ab79b9dd4f11418350/src/utils/getActivityId/index.js#L4) ?

Comment (Shivz3232) [reply src/utils/validateEntity/index.js:4]: Instead of hardcoding...

**Review by Shivz3232 — APPROVED** [review]

File: src/routes/v1/user.route.js [hunk 4]
```diff
@@ -34,4 +34,6 @@ router.post('/activity', insertUserActivitySanitizer, userActivityController.ins
 router.get('/activities', getUserActivitiesSanitizer, userActivityController.getUserActivities);
 router.get('/activity/:activityId', getUserActivitySanitizer, userActivityController.getUserActivity);
 
+router.get('/referral', addUserReferralSanitizer, userController.addUserRefferal);
```
Comment (Shivz3232) [inline src/routes/v1/user.route.js]: Should be a POST request

File: src/controllers/user.controller/addUserRefferal/index.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      success: false,
+      errorCode: 'UserNotFound',
+      errorMessage: 'No user found with this email id',
+    });
+  }
+
+  const referrerId = checkIfUserExistsResponse.result.data.userWithEmail[0].id;
+  const userId = checkIfUserExistsResponse.result.data.userWithCognitoSub[0].id;
+
+  const checkIfReferalExistsResponse = await Hasura(checkIfReferalExists, {
```
Comment (Shivz3232) [inline src/controllers/user.controller/addUserRefferal/index.js]: Hey, I checked this queries, you are checking if current user has been referred from the provided email. But we should reject if the current user has been referred even once, that is even with other email. Ex: If shivu has already referred Pratik, sarang can't refer Pratik.

Comment (Shivz3232) [inline src/controllers/user.controller/addUserRefferal/index.js]: Also lets move this check before getUserDetails. We can avoid one query if this check fails.

**Review by Shivz3232 — CHANGES_REQUESTED** [review]

File: src/controllers/user.controller/addUserRefferal/index.js (hunk 5 above)
Comment (C0SMlC) [reply src/controllers/user.controller/addUserRefferal/index.js]: Yes

File: src/controllers/user.controller/addUserRefferal/index.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      success: false,
+      errorCode: 'UserNotFound',
+      errorMessage: 'No user found with this email id',
+    });
+  }
+
+  const referrerId = checkIfUserExistsResponse.result.data.userWithEmail[0].id;
+  const userId = checkIfUserExistsResponse.result.data.userWithCognitoSub[0].id;
+
+  const addUserReferralResponse = await Hasura(addUserReferral, {
```
Comment (Shivz3232) [inline src/controllers/user.controller/addUserRefferal/index.js:64]: What if the current user inputs his own email? He should not be able to refer himself. Please add the check & see if you are missing any other edge cases.

**Review by Shivz3232 — CHANGES_REQUESTED** [review]

**Pull request merged**